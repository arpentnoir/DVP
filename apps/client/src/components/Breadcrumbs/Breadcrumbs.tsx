import { capitaliseFirstLetter, constructPath } from '@dvp/vc-ui';
import ArrowForwardIcon from '@mui/icons-material/ArrowForward';
import { Box, Breadcrumbs as MuiBreadcrumbs, Link } from '@mui/material';
import useTheme from '@mui/material/styles/useTheme';
import { FunctionComponent } from 'react';
import { Link as RouterLink, useLocation } from 'react-router-dom';

export const Breadcrumbs: FunctionComponent = () => {
  const { pathname } = useLocation();
  const theme = useTheme();

  const routes = pathname.split('/').filter(Boolean);
  return (
    <>
      {routes.length > 0 && (
        <div role="presentation">
          <MuiBreadcrumbs
            aria-label="breadcrumb"
            sx={{
              display: 'flex',
              justifyContent: 'center',
              height: '42px',
              background: theme.palette.common.neutralGrey,
              fontSize: '15px',
              'li:nth-last-child(2)': { transform: 'rotate(45deg)' },
              '@media print': { display: 'none' },
            }}
            separator={
              <Box
                data-testid="breadcrumbs-separator"
                sx={{
                  display: 'flex',
                  justifyContent: 'center',
                  paddingTop: '1px',
                }}
              >
                <ArrowForwardIcon
                  sx={{ fontSize: '14px', color: theme.palette.error.main }}
                />
              </Box>
            }
          >
            <Link
              component={RouterLink}
              underline="hover"
              color={theme.palette.common.textDark}
              to={'/'}
            >
              Home
            </Link>
            {[...routes, ''].map((route, index) => {
              if (index !== routes.length - 1) {
                return (
                  <Link
                    component={RouterLink}
                    key={route}
                    underline="hover"
                    color={theme.palette.common.textDark}
                    to={constructPath(routes.slice(0, index + 1))}
                    tabIndex={0}
                  >
                    {capitaliseFirstLetter(route).replace(/-/g, ' ')}
                  </Link>
                );
              }
              return;
            })}
          </MuiBreadcrumbs>
        </div>
      )}
    </>
  );
};
