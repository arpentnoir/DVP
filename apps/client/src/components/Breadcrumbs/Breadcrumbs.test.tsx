import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { Breadcrumbs } from './Breadcrumbs';

describe('Breadcrumbs', () => {
  it('should render correctly', () => {
    const { baseElement } = render(
      <MemoryRouter initialEntries={['/test/route']}>
        <Breadcrumbs />
      </MemoryRouter>
    );

    expect(baseElement).toMatchSnapshot();
  });

  it('should display links for each subdirectory except for the last', () => {
    const { queryByText, queryAllByTestId } = render(
      <MemoryRouter initialEntries={['/test/breadcrumbs/path']}>
        <Breadcrumbs />
      </MemoryRouter>
    );

    const separators = queryAllByTestId('breadcrumbs-separator');

    expect(queryByText('Home')).toBeInTheDocument();
    expect(queryByText('Home')?.closest('a')).toHaveAttribute('href', '/');

    expect(queryByText('Test')).toBeInTheDocument();
    expect(queryByText('Test')?.closest('a')).toHaveAttribute('href', '/test');

    expect(queryByText('Breadcrumbs')).toBeInTheDocument();
    expect(queryByText('Breadcrumbs')?.closest('a')).toHaveAttribute(
      'href',
      '/test/breadcrumbs'
    );

    expect(separators.length).toBe(3);
    expect(separators[separators.length - 1].parentElement).toHaveStyle(
      'transform: rotate(45deg)'
    );

    expect(queryByText('Path')).not.toBeInTheDocument();
  });

  it('should not display breadcrumbs if on the homepage', () => {
    const { queryByText, queryAllByTestId } = render(
      <MemoryRouter initialEntries={['/']}>
        <Breadcrumbs />
      </MemoryRouter>
    );

    expect(queryByText('Home')).not.toBeInTheDocument();
    expect(queryAllByTestId('breadcrumbs-separator').length).toBe(0);
  });

  it('should only display home link for 1 level deep', () => {
    const { queryByText, queryAllByTestId } = render(
      <MemoryRouter initialEntries={['/test']}>
        <Breadcrumbs />
      </MemoryRouter>
    );

    const separators = queryAllByTestId('breadcrumbs-separator');

    expect(queryByText('Home')).toBeInTheDocument();
    expect(queryByText('Home')?.closest('a')).toHaveAttribute('href', '/');

    expect(queryByText('Test')).not.toBeInTheDocument();

    expect(separators.length).toBe(1);
    expect(separators[separators.length - 1].parentElement).toHaveStyle(
      'transform: rotate(45deg)'
    );
  });

  it('should not display a link for the current page', () => {
    const { queryByText } = render(
      <MemoryRouter initialEntries={['/test/documents']}>
        <Breadcrumbs />
      </MemoryRouter>
    );

    expect(queryByText('Documents')).not.toBeInTheDocument();
  });

  it('should display a rotated separator as the last element', () => {
    const { queryAllByTestId } = render(
      <MemoryRouter initialEntries={['/test/documents']}>
        <Breadcrumbs />
      </MemoryRouter>
    );

    const separators = queryAllByTestId('breadcrumbs-separator');

    expect(separators.length).toBe(2);
    expect(separators[separators.length - 1].parentElement).toHaveStyle(
      'transform: rotate(45deg)'
    );
  });
});
