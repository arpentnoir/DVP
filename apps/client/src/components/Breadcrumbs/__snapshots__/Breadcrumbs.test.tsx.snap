// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumbs should render correctly 1`] = `
<body>
  <div>
    <div
      role="presentation"
    >
      <nav
        aria-label="breadcrumb"
        class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-1a3e67b-MuiTypography-root-MuiBreadcrumbs-root"
      >
        <ol
          class="MuiBreadcrumbs-ol css-4pdmu4-MuiBreadcrumbs-ol"
        >
          <li
            class="MuiBreadcrumbs-li"
          >
            <a
              class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineHover css-17dcqu8-MuiTypography-root-MuiLink-root"
              href="/"
            >
              Home
            </a>
          </li>
          <li
            aria-hidden="true"
            class="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
          >
            <div
              class="MuiBox-root css-69knei"
              data-testid="breadcrumbs-separator"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1nhaepu-MuiSvgIcon-root"
                data-testid="ArrowForwardIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="m12 4-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8z"
                />
              </svg>
            </div>
          </li>
          <li
            class="MuiBreadcrumbs-li"
          >
            <a
              class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineHover css-17dcqu8-MuiTypography-root-MuiLink-root"
              href="/test"
              tabindex="0"
            >
              Test
            </a>
          </li>
          <li
            aria-hidden="true"
            class="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
          >
            <div
              class="MuiBox-root css-69knei"
              data-testid="breadcrumbs-separator"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1nhaepu-MuiSvgIcon-root"
                data-testid="ArrowForwardIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="m12 4-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8z"
                />
              </svg>
            </div>
          </li>
          <li
            class="MuiBreadcrumbs-li"
          >
            <a
              class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineHover css-17dcqu8-MuiTypography-root-MuiLink-root"
              href="/test/route"
              tabindex="0"
            />
          </li>
        </ol>
      </nav>
    </div>
  </div>
</body>
`;
